### PR TITLE
nautilus: rbd-mirror: image replayer stop might race with instance replayer shut down

### PR DESCRIPTION
**Pull request by trociny**
backport tracker: https://tracker.ceph.com/issues/45275
backport tracker: https://tracker.ceph.com/issues/45764

---

backport of https://github.com/ceph/ceph/pull/34615
parent tracker: https://tracker.ceph.com/issues/45072

backport of https://github.com/ceph/ceph/pull/3493
parent tracker: https://tracker.ceph.com/issues/45716

this backport was staged using ceph-backport.sh version 16.0.0.6848
find the latest version at https://github.com/ceph/ceph/blob/master/src/script/ceph-backport.sh

### CODE REVIEW

**Comment by trociny** [comment]
Cherry-picking 64f8d9c30c53077f48107e83ba836897aa99e609 was skipped because it adds changes (switch to common C_TrackedOp context class) to the code that does not exist in nautilus.

**Comment by yuriw** [comment]
test this please

**Comment by idryomov** [comment]
What about follow-up fixes in https://github.com/ceph/ceph/pull/34931?  Are they needed?

**Comment by trociny** [comment]
> What about follow-up fixes in #34931? Are they needed?

Sure. I was going to add them here after making sure this part passes jenkins test.

**Comment by trociny** [comment]
The backport of  #34931 is added.

File: src/tools/rbd_mirror/ImageReplayer.cc [hunk 1]
```diff
@@ -842,14 +853,14 @@ void ImageReplayer<I>::handle_replay_ready()
 template <typename I>
 void ImageReplayer<I>::restart(Context *on_finish)
 {
-  FunctionContext *ctx = new FunctionContext(
+  auto ctx = new FunctionContext(
```
Comment (idryomov) [inline src/tools/rbd_mirror/ImageReplayer.cc:856]: Nit: unnecessary change

File: src/tools/rbd_mirror/ImageReplayer.cc [hunk 2]
```diff
@@ -842,14 +853,14 @@ void ImageReplayer<I>::handle_replay_ready()
 template <typename I>
 void ImageReplayer<I>::restart(Context *on_finish)
 {
```
Comment (idryomov) [inline src/tools/rbd_mirror/ImageReplayer.cc:855]: The original commit adds setting of `m_restart_requested` under `m_lock` here.  Why is it omitted?

Comment (trociny) [reply src/tools/rbd_mirror/ImageReplayer.cc:855]: Oh, sorry. I lost this when resolving the conflict. Should be ok now.

**Comment by trociny** [comment]
@idryomov Thanks. Updated

**Review by idryomov — APPROVED** [review]

**Comment by idryomov** [comment]
http://qa-proxy.ceph.com/teuthology/yuriw-2021-06-14_16:00:31-rbd-wip-yuri-testing-2021-06-14-0729-nautilus-distro-basic-smithi/6171891/teuthology.log
```
2021-06-14T20:52:53.399 INFO:tasks.rbd_mirror.cluster2.client.mirror.2.smithi174.stderr:*** Caught signal (Aborted) **
2021-06-14T20:52:53.399 INFO:tasks.rbd_mirror.cluster2.client.mirror.2.smithi174.stderr: in thread 7f69788171c0 thread_name:rbd-mirror
2021-06-14T20:52:53.425 INFO:tasks.rbd_mirror.cluster2.client.mirror.2.smithi174.stderr: ceph version 14.2.21-364-gefaec7927e9 (efaec7927e949dc1e9e7f068e4f86265596ffab6) nautilus (stable)
2021-06-14T20:52:53.426 INFO:tasks.rbd_mirror.cluster2.client.mirror.2.smithi174.stderr: 1: (()+0xf630) [0x7f696d7f1630]
2021-06-14T20:52:53.426 INFO:tasks.rbd_mirror.cluster2.client.mirror.2.smithi174.stderr: 2: (pthread_cond_wait()+0xc5) [0x7f696d7eda35]
2021-06-14T20:52:53.426 INFO:tasks.rbd_mirror.cluster2.client.mirror.2.smithi174.stderr: 3: (rbd::mirror::InstanceReplayer<librbd::ImageCtx>::shut_down()+0x197) [0x55d59d86ec77]
2021-06-14T20:52:53.427 INFO:tasks.rbd_mirror.cluster2.client.mirror.2.smithi174.stderr: 4: (rbd::mirror::PoolReplayer<librbd::ImageCtx>::shut_down()+0x96) [0x55d59d830a06]
2021-06-14T20:52:53.427 INFO:tasks.rbd_mirror.cluster2.client.mirror.2.smithi174.stderr: 5: (rbd::mirror::PoolReplayer<librbd::ImageCtx>::~PoolReplayer()+0xa4) [0x55d59d834cd4]
2021-06-14T20:52:53.427 INFO:tasks.rbd_mirror.cluster2.client.mirror.2.smithi174.stderr: 6: (std::_Rb_tree<std::pair<long, rbd::mirror::PeerSpec>, std::pair<std::pair<long, rbd::mirror::PeerSpec> const, std::unique_ptr<rbd::mirror::PoolReplayer<librbd::ImageCtx>, std::default_delete<rbd::mirror::PoolReplayer<librbd::ImageCtx> > > >, std::_Select1st<std::pair<std::pair<long, rbd::mirror::PeerSpec> const, std::unique_ptr<rbd::mirror::PoolReplayer<librbd::ImageCtx>, std::default_delete<rbd::mirror::PoolReplayer<librbd::ImageCtx> > > > >, std::less<std::pair<long, rbd::mirror::PeerSpec> >, std::allocator<std::pair<std::pair<long, rbd::mirror::PeerSpec> const, std::unique_ptr<rbd::mirror::PoolReplayer<librbd::ImageCtx>, std::default_delete<rbd::mirror::PoolReplayer<librbd::ImageCtx> > > > > >::_M_erase(std::_Rb_tree_node<std::pair<std::pair<long, rbd::mirror::PeerSpec> const, std::unique_ptr<rbd::mirror::PoolReplayer<librbd::ImageCtx>, std::default_delete<rbd::mirror::PoolReplayer<librbd::ImageCtx> > > > >*)+0x3f) [0x55d59d82bfef]
2021-06-14T20:52:53.427 INFO:tasks.rbd_mirror.cluster2.client.mirror.2.smithi174.stderr: 7: (rbd::mirror::Mirror::~Mirror()+0xaf) [0x55d59d8278df]
2021-06-14T20:52:53.428 INFO:tasks.rbd_mirror.cluster2.client.mirror.2.smithi174.stderr: 8: (main()+0x3a8) [0x55d59d8171c8]
2021-06-14T20:52:53.428 INFO:tasks.rbd_mirror.cluster2.client.mirror.2.smithi174.stderr: 9: (__libc_start_main()+0xf5) [0x7f696bd94555]
2021-06-14T20:52:53.428 INFO:tasks.rbd_mirror.cluster2.client.mirror.2.smithi174.stderr: 10: (()+0x215090) [0x55d59d824090]
2021-06-14T20:52:53.430 INFO:tasks.rbd_mirror.cluster2.client.mirror.2.smithi174.stderr:2021-06-14 20:52:53.430 7f69788171c0 -1 *** Caught signal (Aborted) **
2021-06-14T20:52:53.430 INFO:tasks.rbd_mirror.cluster2.client.mirror.2.smithi174.stderr: in thread 7f69788171c0 thread_name:rbd-mirror
```

**Comment by idryomov** [comment]
I commented here thinking `rbd-mirror` crashed during shutdown, but it actually hung and was aborted on purpose.  Rerun expectedly succeeded.

@trociny Can you take a look to make sure the hang is not related?  Apart from this PR, https://github.com/ceph/ceph/pull/41787 and https://github.com/ceph/ceph/pull/41788 were included.

**Comment by trociny** [comment]
The hang looks related. It hanged in `InstanceReplayer` on shutdown, "waiting for in-flight start/stop/restart", i.e. exactly at the backported code. Right now I have no idea about the cause but I am still looking.

**Comment by trociny** [comment]
> The hang looks related. It hanged in `InstanceReplayer` on shutdown, "waiting for in-flight start/stop/restart", i.e. exactly at the backported code. Right now I have no idea about the cause but I am still looking.

Actually, the hang may be not related (it might have been waiting for some other tracked op, not one that was added in this PR). But still it looks highly suspicious. And I have failed to track why it could get stuck so far.

I am going to continue to investigate this tomorrow when I have time. But if we are shot in time (for release) I think we can just exclude this PR from the release (not closing, probably merging after the release). The initial issue does not look critical -- the assertion could fail on the rbd-mirror shutdown, so users probably not even notice it if they hit it.

**Comment by yuriw** [comment]
@trociny any updates on this ?

**Comment by trociny** [comment]
I tracked the issue to the bug in `ImageReplayer<I>handle_start_replay`. If we cancel an image replay start when it is in `start_replay`, then in `handle_start_replay` the `on_replay_interrupted()` [2] will return true and we just return, without completing `m_on_start_finish` context, so this "start" remains pending in tracked operations and on shut down we are waiting for it to complete forever.

So it is not a bug in the backport but it reveals a bug in nautilus. In the newer versions `ImageReplayer<I>handle_start_replay` does not has this bug (we call `on_start_interrupted()` as the first step, which handles this case). Now I am trying to figure out if there is a commit that I just can cherry-pick or if it should be a direct fix to nautilus.

[1] https://github.com/ceph/ceph/blob/nautilus/src/tools/rbd_mirror/ImageReplayer.cc#L635
[2] https://github.com/ceph/ceph/blob/nautilus/src/tools/rbd_mirror/ImageReplayer.cc#L664
[3] https://github.com/ceph/ceph/blob/master/src/tools/rbd_mirror/ImageReplayer.cc#L427

**Comment by trociny** [comment]
I believe it was incidentally fixed during refactoring in 0d36eb58f8d5625bad0c075632594ae5e1169fa2, which we can't backport. So it should be a direct bug fix commit.

**Comment by yuriw** [comment]
@trociny I will start 14.2.22 testing and we will decide on this when you are ready, thx !

File: src/tools/rbd_mirror/ImageReplayer.cc [hunk 3]
```diff
@@ -635,7 +639,9 @@ template <typename I>
 void ImageReplayer<I>::handle_start_replay(int r) {
   dout(10) << "r=" << r << dendl;
 
-  if (r < 0) {
+  if (on_start_interrupted()) {
+    return;
+  } if (r < 0) {
```
Comment (idryomov) [inline src/tools/rbd_mirror/ImageReplayer.cc:644]: This happens to work because of `return`, but better to change to `else if`.

**Comment by idryomov** [comment]
jenkins test make check

File: src/tools/rbd_mirror/ImageReplayer.cc (hunk 3 above)
Comment (trociny) [reply src/tools/rbd_mirror/ImageReplayer.cc:644]: Ah. Sure! Thanks!

**Comment by trociny** [comment]
Updated.

Here are teuthology results for a test subset (`--filter rbd/mirror-thrash --limit 20`) [1] (no failures). It is with the previous patch though it should not make any difference. For comparison the results for the same subset without the patch [2] (some tests failed).

[1] https://pulpito.ceph.com/trociny-2021-06-17_16:39:29-rbd-wip-mgolub-testing-nautilus-distro-basic-smithi/
[2] https://pulpito.ceph.com/trociny-2021-06-17_08:21:42-rbd-wip-mgolub-testing-nautilus-distro-basic-smithi/

**Review by idryomov — APPROVED** [review]
